scarlet/bbox.py

```diff
@@ -293,18 +223,12 @@ def __and__(self, other):
             The rectangular box that is in the overlap region
             of both boxes.
         """
-        front = max(self.front, other.front)
-        back = min(self.back, other.back)
-        bottom = max(self.bottom, other.bottom)
-        top = min(self.top, other.top)
-        left = max(self.left, other.left)
-        right = min(self.right, other.right)
-        return Box.from_bounds(front, back, bottom, top, left, right)
-
-    def __str__(self):
-        return "Box({0}..{1}, {2}..{3}, {4}..{5})".format(
-            self.front, self.back, self.bottom, self.top, self.left, self.right
-        )
+        assert other.D == self.D
+        bounds = []
+        for d in range(self.D):
+            bounds.append(max(self.start[d], other.start[d]))
+            bounds.append(min(self.stop[d], other.stop[d]))
+        return Box.from_bounds(*bounds)
```

Comment:
We should check that the minimum bound is less than the maximum bound. It might be that the sources don't overlap at all, in which case we should either raise an `Exception`, return `None`, or return a box with size `(0,)*D`. My preference would be for the latter option except that it isn't clear where to put origin. Suggestions welcome.

Reply:
Solved it by checking that bounds are ordered (min, max), which would be violated for non-overlapping boxes.

Reply:
What does it do if they are not ordered?

Reply:
shape = 0

Reply:
Ok, so looking at the code if I have one box from `(0,0)` to `(1,1)` and another from `(0,2)` to `(2,3)` it looks like the result of `box1 & box2` would be a new box with origin `(0,2)` and `shape=(1,0)`. This isn't really the overlap, since there is no overlap. In other words the shape should be `shape=(0,0)`. And, as I mentioned above, this leaves us with a decision to make about where to put the origin of an empty box, since the concept of an origin doesn't really make sense here. If we decide that we should return an empty box, then I would want the return to be `Box((0,)*self.D, origin=?)`.

I still don't like the idea of an empty box having an origin unless it's something like `(np.inf, np.inf)` or `None`, since it could be confusing to set `overlap = (box1 & box2)` and then go on performing calculations with `overlap.origin`. So I'm starting to lean more towards returning `None` if there is no overlap or inherit a new `EmptyBox` class from `Box` where `EmptyBox` defines a box with `shape=(0,0)` and `origin=None`.

Reply:
I disagree with both paragraphs:
* since the logic behind the nD code is that boxes are n intervals, one for each dimension, the logic of your overlap example is correct: there is overlap in the first dimension, but not in the last. this means that the overlap has no volume, and no point is contained. I don't see any relevant arithmetic that can be flawed because the Box has some non-zero shape and an arbitrary origin. In particular, the slice of this box is empty for any array given. That's the function the BBox needs to implement.
* Your suggestion necessitates extra code and checking of corner cases. Mine doesn't.

Reply:
I get your point, in that if we project the boxes onto the first dimension then there would be an overlap, but there is still an inconsistency. For example, use the overlap box described above, namely
```
overlap = Box(shape=(1,0), origin=(0,2))
```
Now calculate `overlap.contains((1,2))`, which results in `True`! Of course had we chosen a different coordinate for the arbitrary (non-overlapping) dimension, for example `origin=(0,0)` then `overlap.contains((0,0))` would return `True` and all other points would return `False`. So no matter where we put the origin, as long as it's a finite number then there will always be one point contained in a non-overallping `Box` that your code returns, and it's an arbitrary point.

I think that we can all agree that no points should be contained in a box with zero volume, right? So at the very least we need to add a check in `contains` that all of the dimensions are non-zero, but preferably if you want to allow for overlapping projections then I would want the origin to be `(0, None)` or `(0,np.inf)` to prevent improper interpretation of the origin.

Reply:
```
overlap = scarlet.Box(shape=(1,0), origin=(0,2))
overlap.contains((1,2))
```
yields `False`, as it should.

Reply:
Same for `overlap.contains(overlap.origin)` or any other point

Reply:
Ah, I didn't notice that you changed that line in `contains` this morning. So while I don't like this behavior, I think that I won't like whatever we come up with because I don't think that there is a good solution in terms of what we call the origin in the non-overlapping dimension. So I'm willing to accept this for now with the caveat that we might need to revisit this later.

Reply:
Yeah, I fixed that logic bug in `contains` so that everything is consistent with python slicing. For what it's worth, I also don't *like* a dangling `origin`, but I just don't think it's worth extra code.